internal/venice-common/src/main/java/com/linkedin/venice/acl/handler/AbstractStoreAclHandler.java

```diff
@@ -55,14 +118,18 @@ public AbstractStoreAclHandler(
    */
   @Override
   public void channelRead0(ChannelHandlerContext ctx, HttpRequest req) throws SSLPeerUnverifiedException {
-    if (isAccessAlreadyApproved(ctx)) {
+    Channel originalChannel = ServerHandlerUtils.getOriginalChannel(ctx);
+    if (originalChannel == null) {
+      NettyUtils.setupResponseAndFlush(HttpResponseStatus.BAD_REQUEST, BAD_REQUEST_RESPONSE, false, ctx);
+      return;
+    }
+    if (isAccessAlreadyApproved(originalChannel)) {
       ReferenceCountUtil.retain(req);
       ctx.fireChannelRead(req);
       return;
     }
 
     String uri = req.uri();
-
     // Parse resource type and store name
     String[] requestParts = URI.create(uri).getPath().split("/");
```

Comment:
Unrelated to this PR so feel free to ignore, but... the parsing of the URI into request parts happens in several handlers... it may be good to have some utility function that can cache it into the channel (though it would need to be on a per-request basis only, not longer-lived than that; we do a similar thing for when we extract the key count in the router, IIRC). Then we would do the request parsing just once in the whole request.

Reply:
This will be a little tricky as we don't maintain a shared object across all the handlers until `RouterRequestHttpHandler`.
I will leave such optimization outside of this PR.